File: [0098][Medium][Validate Binary Search Tree]/Validate_Binary_Search_Tree_3.cpp
#include <stdlib.h>
using namespace std;
/*   Definition for a binary tree node. */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}

};
// leetcode time     cost : 28 ms
// leetcode memory   cost : 20.7 MB
//C++ inorder recursion:
class Solution {
public:
    TreeNode* pre;
    bool isValidBST(TreeNode* root) {        
        //solution with recursion.
        if(root == NULL)
            return true;
        if(!isValidBST(root->left)){
            return false;
        }
        if(pre && pre->val >= root->val)
            return false;
        pre = root;
        if(!isValidBST(root->right))
            return false;
        return true;            
    }
};

File: [0098][Medium][Validate Binary Search Tree]/Validate_Binary_Search_Tree_3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 
// leetcode time     cost : 2 ms
// leetcode memory   cost : 37 MB
//inorder traversal, check result is a  increasing list or not. 
class Solution {
    public boolean isValidBST(TreeNode root) {
        // use the inorder traversal of BST to judge it is a increasing order list or not.
        List<Integer> list = new ArrayList<>();
        inOrder(root, list);
        for (int i = 1; i < list.size(); i++) {
            if (list.get(i - 1) >= list.get(i)) {
                return false;
            }
        }
        return true;
    }

    private void inOrder(TreeNode node, List<Integer> list) {
        if (node == null) {
            return;
        }
        inOrder(node.left, list);
        list.add(node.val);
        inOrder(node.right, list);
    }
}